### PR TITLE
Proposal/Feature add flag --skip-empty-annotation to honor only annotated resources.

### PR DESCRIPTION
**Pull request by nutellinoit**
**What this PR does / why we need it**:
This PR addresses the need to limit kong ingress controller on only the annotated resources. Mainly when using multiple ingress controller in a Kubernetes cluster.

**Special notes for your reviewer**:
I added only a flag on the ingress controller `--skip-empty-annotation` that when set to `true` skips the evaluation of all non annotated resources.

I tried the setup on minikube with some ingresses and kong CRDs and everything is working. Also `make test-all` is passing and I added some new test case on `internal/ingress/annotations/annotations_test.go` test suite.

Thank you!

### CODE REVIEW

**Comment by hbagdi** [comment]
We are discussing this internally if this should be added or not. This is something that has been asked for quite a bit. We need to figure out what would be a good default value and what would be the recommendation going forward, especially with Ingress V1 in Kubernetes 1.18 which changes how class should be interpreted. We will look into this post 0.9 release which is scheduled for next week.

**Comment by nutellinoit** [comment]
Thank you @hbagdi !

File: docs/concepts/deployment.md [hunk 1]
```diff
@@ -264,6 +264,7 @@ There are a few different ways of accomplishing this:
   where any ingress resource that is not annotated is picked up.
   Therefore with different ingress class then `kong`, you have to use that
   ingress class with every Kong CRD object (plugin, consumer) which you use.
+  You can force the honoring of only `kong` ingress class using the flag `--skip-empty-annotation=true`.
```
Comment (hbagdi) [inline docs/concepts/deployment.md:267]: Please remove the doc update from this PR. We will do doc updates later.

File: go.mod [hunk 2]
```diff
@@ -26,7 +26,9 @@ require (
 	github.com/tidwall/gjson v1.2.1
 	github.com/tidwall/match v1.0.1 // indirect
 	github.com/tidwall/pretty v0.0.0-20190325153808-1166b9ac2b65 // indirect
+	golang.org/x/lint v0.0.0-20200302205851-738671d3881b // indirect
```
Comment (hbagdi) [inline go.mod:29]: These dependencies shouldn't be added. Please clean go.mod and go.sum files.

File: internal/ingress/annotations/annotations.go [hunk 3]
```diff
@@ -46,15 +46,15 @@ const (
 	DefaultIngressClass = "kong"
 )
 
-func validIngress(ingressAnnotationValue, ingressClass string) bool {
+func validIngress(ingressAnnotationValue, ingressClass string, skipEmptyAnnotation bool) bool {
```
Comment (hbagdi) [inline internal/ingress/annotations/annotations.go:49]: Can you rename the variable to `skipClasslessIngress`?
Ingress class is a formalized concept in Ingress v1 and we would like to reuse parts of this PR for that as well.

File: cli/ingress-controller/flags.go [hunk 4]
```diff
@@ -158,6 +159,8 @@ Kong's Admin SSL certificate.`)
 	// Resource filtering
 	flags.String("watch-namespace", apiv1.NamespaceAll,
 		`Namespace to watch for Ingress. Default is to watch all namespaces`)
+	flags.Bool("skip-empty-annotation", false,
```
Comment (hbagdi) [inline cli/ingress-controller/flags.go:162]: Please rename this flag to `skip-classless-ingress-v1beta1`.
We plan to introduce another flag to control this behavior for Ingress v1 as well.

File: cli/ingress-controller/flags.go [hunk 5]
```diff
@@ -53,9 +53,10 @@ type cliConfig struct {
 	KongAdminCACertPath    string
 
 	// Resource filtering
-	WatchNamespace string
-	IngressClass   string
-	ElectionID     string
+	WatchNamespace      string
+	SkipEmptyAnnotation bool
```
Comment (hbagdi) [inline cli/ingress-controller/flags.go:57]: Please rename this field to `SkipClasslessIngressV1beta1`. Please rename other fields accordingly.

File: internal/ingress/annotations/annotations.go [hunk 6]
```diff
@@ -46,15 +46,15 @@ const (
...
 	// we have 2 valid combinations
 	// 1 - ingress with default class | blank annotation on ingress
 	// 2 - ingress with specific class | same annotation on ingress
 	//
 	// and 2 invalid combinations
 	// 3 - ingress with default class | fixed annotation on ingress
 	// 4 - ingress with specific class | different annotation on ingress
-	if ingressAnnotationValue == "" && ingressClass == DefaultIngressClass {
+	if ingressAnnotationValue == "" && !skipEmptyAnnotation && ingressClass == DefaultIngressClass {
```
Comment (hbagdi) [inline internal/ingress/annotations/annotations.go:57]: Can you please update the code comments according to this change?
This is a tricky logic and we should document it for future-selves.

File: internal/ingress/annotations/annotations_test.go [hunk 7]
```diff
@@ -27,16 +27,20 @@ import (
 
 func TestIngressClassValidatorFunc(t *testing.T) {
 	tests := []struct {
-		ingress    string
-		controller string
-		isValid    bool
+		ingress      string
+		forceIngress bool
```
Comment (hbagdi) [inline internal/ingress/annotations/annotations_test.go:31]: Please rename `forceIngress` to something meaningful.
How about `SkipClasslessIngress`?

File: internal/ingress/annotations/annotations_test.go [hunk 8]
```diff
@@ -27,16 +27,20 @@ import (
...
-		{"custom", "kong", false},
+		{"", false, "", true},
+		{"", false, "kong", true},
+		{"", true, "kong", false},
+		{"kong", false, "kong", true},
+		{"kong", true, "kong", true},
+		{"custom", false, "custom", true},
+		{"", false, "killer", false},
+		{"custom", false, "kong", false},
+		{"custom", true, "kong", false},
```
Comment (hbagdi) [inline internal/ingress/annotations/annotations_test.go:43]: Can we add `{"", true, "custom", false}`?

File: internal/ingress/store/store.go [hunk 9]
```diff
@@ -102,11 +104,12 @@ type CacheStores struct {
 }
 
 // New creates a new object store to be used in the ingress controller
-func New(cs CacheStores, ingressClass string) Storer {
+func New(cs CacheStores, ingressClass string, skipEmptyAnnotation bool) Storer {
```
Comment (hbagdi) [inline internal/ingress/store/store.go:107]: This needs to be refactored away but let's keep that out of scope for this PR.

**Comment by hbagdi** [comment]
Also, please rebase on latest `next` branch.

**Comment by nutellinoit** [comment]
Hi @hbagdi , i will make the requested changes as soon as possible, thank you for the feedback!

**Comment by nutellinoit** [comment]
Hi @hbagdi , i did all the requested changes!

File: internal/ingress/store/fake_store.go [hunk 10]
```diff
@@ -140,7 +140,7 @@ func NewFakeStore(
 
 			KnativeIngress: knativeIngressStore,
 		},
-		isValidIngresClass: annotations.IngressClassValidatorFuncFromObjectMeta("kong"),
+		isValidIngresClass: annotations.IngressClassValidatorFuncFromObjectMeta("kong", false),
```
Comment (hbagdi) [inline internal/ingress/store/fake_store.go:143]: Can you make this configurable and tweak `NewFakeStore` to pass this in?

Comment (nutellinoit) [reply internal/ingress/store/fake_store.go:143]: Sure!

**Comment by nutellinoit** [comment]
> Friendly ping on an update on this.

I will look into it this week! Thanks for the ping

**Comment by nutellinoit** [comment]
Hi @hbagdi , i did the requested changes!

**Comment by hbagdi** [comment]
Thank you for your contribution. Please fill out the following form to claim your contributor swag:
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributor-t-shirt.

**Pull request merged**